python/GafferUI/NodeSetEditor.py

```diff
@@ -61,22 +77,149 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
+	# If drivingEditor is None, any existing links will be broken.
+	def setNodeSetDriver( self, drivingEditor, mode = DriverModeNodeSet ) :
+
+		if drivingEditor is not None :
+			assert( isinstance( drivingEditor, GafferUI.NodeSetEditor ) )
+			# We also need to stop people creating infinite loops
+			if GafferUI.NodeSetEditorAlgo.drivesNodeSet( self, drivingEditor ) :
+				raise ValueError( "The supplied driver is already driven by this editor" )
+
+		if self.__nodeSetDriver :
+			previousDriver = self.__nodeSetDriver["weakDriver"]()
+			# It may have died, so we sill will have data, but the ref wil be dead
+			if previousDriver is not None :
+				if drivingEditor is previousDriver and mode == self.__nodeSetDriver["mode"] :
+					return
+				else :
+					previousDriver.__unregisterDrivenEditor( self )
+
+		self.__nodeSetDriver = {}
+
+		if drivingEditor :
+
+			drivingEditor.__registerDrivenEditor( self, mode )
+
+			weakDriver = weakref.ref(
+				drivingEditor,
+				# We need to unlink ourselves if the driver goes away
+				lambda _ : self.setNodeSetDriver( None )
+			)
+			changeCallback = self.__nodeSetDriverModes[ mode ]
+
+			def updateFromDriver( _ ) :
+				if weakDriver() is not None :
+					nodeSet = weakDriver().getNodeSet()
+					if changeCallback :
+						nodeSet = changeCallback( self, weakDriver() )
+					# We need to be able to track un-pinning, so we need to see if someone
+					# else has changed the nodeSet of the editor by some other means, so
+					# keep a copy here.
+					self.__nodeSetDriver["lastNodeSet"] = nodeSet
+					self.__setNodeSetInternal( nodeSet, callUpdateFromSet=True )
+
+			self.__nodeSetDriver = {
+				"mode" : mode,
+				"weakDriver" : weakDriver,
+				"driverNodeSetChangedConnection" : drivingEditor.nodeSetChangedSignal().connect( updateFromDriver ),
+			}
+			updateFromDriver( drivingEditor )
+
+		self.__nodeSetDriverChangedSignal( self )
+
+	## Returns a tuple of the drivingEditor and the drive mode.
+	# When there is no driver ( None, "" ) will be returned.
+	def getNodeSetDriver( self ) :
+
+		if self.__nodeSetDriver :
+			return ( self.__nodeSetDriver["weakDriver"](), self.__nodeSetDriver["mode"] )
+
+		return ( None, "" )
+
+	## Called whenever the editor's driving node set changes.
+	# Note: This is called after nodeSetChangedSignal in the event that
+	# the setNodeSet call breaks an existing driver link.
+	def nodeSetDriverChangedSignal( self ) :
+
+		return self.__nodeSetDriverChangedSignal
+
+	## Returns a dict of { editor : mode } that are driven by this editor.
+	# If recurse is true, the link chain will be followed recursively to
+	# also include editors that indirectly driven by this one.
+	def drivenNodeSets( self, recurse = False ) :
+
+		# Unwrap the weak refs
+		driven = { w(): m for w,m in self.__drivenNodeSets.items() if w() is not None }
+
+		if recurse :
+			distant = {}
+			for editor in driven :
+				distant.update( editor.drivenNodeSets( recurse = True ) )
```

Comment:
Is this two-stage update using `distant` any different than just calling `driven.update()` directly?

Reply:
https://github.com/GafferHQ/gaffer/pull/3270/commits/07eba8b8f359b989d1f79a1a1c7a72a189cc57eb